### ISSUE TITLE
token balances output inconsistency

### ISSUE THREAD

**Issue by buhtignew** [slimcoin-project/pacli#102]
My `token balances -w` output is as following:
```
+Balances of addresses with labels in wallet:--------------+------------+-------------+
| Label               | Address                            | PoB tokens | dPoD tokens |
+---------------------+------------------------------------+------------+-------------+
| donor_address7      | mskxtnWBJ1bBYZxmHe8rTR5H8ayZEXvG35 | 142690.0   | 0           |
| donor_address9      | mfq11jH49F2jqUMrVpGZhtyeDC88Vk9Vs8 | 2400.0     | 0           |
| donor_address8      | mg59ZLqztR5zWQB9HknPNHFH4PHwcq4Ee2 | 0          | 0           |
| origin_address7     | mrnrAyYBG9gNzdAiZ6aRDC3tBkrCsjTnSi | 100.0      | 0           |
| donor_address6      | n1xC4Dkp48mbsJsMV3yYEho6adBmbjgqNV | 10.0       | 0           |
| origin_address4     | n3MtPoPaAREU5GEhAErBPuju83bVog2opz | 1850.0     | 0           |
| donor_address1.2.1  | mskwPWmHNAV2yfKPDCs4Yt6WvSpfGbA64e | 0          | 0           |
| donor_address0.2    | n4AoPomtAQRxsTc2TDMFMsP2PonhZ2vVAz | 4.0        | 0           |
| donor_address3.2    | mpKx1QBBFSPY3b7CjrR4Mvcq4t1tHiAEcz | 20.0       | 0           |
| donor_address1.2    | mzHSKpCYK2J4Np6cVimPQsJufU7ZwesrcL | 0          | 0           |
| origin_address5     | mjXqewNUxUCW4v6bQ7CzhRAVMLHmQ6ktxN | 2100.0     | 0           |
| donor_address5      | mq1jZAiu8NoL3U3Ed7YiDdfjmY5t6Zphvq | 9999.0     | 0           |
| donor_address4      | mrnZpB5Wm1FDBBcZ1s1pdPHpQgzaLRteSr | 2000.0     | 0           |
| donor_address3      | mfemhfkdD9tZG7Lukv2dT7vdq8d5xYFQCX | 1900.0     | 0           |
| donor_address1.1    | mhg25VziknSgQLBg6uCMBXsT8VpwbaQUJR | 10.0       | 0           |
| donor_address2      | mwgpxrjnzXxfXSdZaD6WvtH1122B1fwMv4 | 1800.0     | 0           |
| donor_address       | mukgdQJzBzCkMVjmUNgcmDJD5cwY3zzua6 | 1.0        | 0           |
| origin_address2     | mfkwJpC1RaEAPrh6bNk849jFyzMgBA3td1 | 0          | 0           |
| origin_address1     | mxuqE6UWREDuWuMkBWoz1F8jfZur5CM3ZH | 1700.0     | 0           |
| main                | mw6J8BWwsbERhCmVYPdS4Kk5W21FuK3Qt3 | 1.0        | 0           |
| new_one             | mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya | 1209.0     | 9.0         |
| old_default         | n2nSbZYLV77NFHSD3P4ZdFKyp65rtTHP1b | 102366.0   | 0           |
| donor_address1      | mx5MdsenFDZufFuwT9ND7BKQBzS4Hy9YUP | 73410.0    | 0           |
| super_new           | msAzW8dLyCNhnashwThAo2rUNKBycqt7yv | 0          | 0           |
| test_label          | mgpiP2Dc5QweKFS55HvRqsWyQ6PMXJCVCk | 0          | 0           |
| keyring_address     | mw9gjLtQVhu34YKep7vrZJDiong69Zmyid | 0          | 0           |
| second_one          | mjAykGXoJu21VZLr5JymuQXFCXyzhgn3Cb | 0          | 0           |
| flag-l              | mgUkRF2eosnFQpMKqAYxuzVLxccTPyNxYw | 0          | 0           |
| no_transactions?    | mqZ5nqs6bpaAqEoEV9jVFfwsM3hBZD4C3J | 0          | 0           |
| locator_test        | mmWqnW9DkPq2Dosm8ikmNvd8xfaeynp6iw | 0          | 0           |
| locator_test1       | myu48LbhEumui6yMkdD5DgYYf7STgEtF7Y | 0          | 0           |
| locator_test2       | mgpKAUYnTyo9FKEhHKYeXMzBFkLdPBuW8j | 0          | 0           |
| new_after_refill    | mwvsHNwNQ8eiRFGHNqnpD7EdoFk5zzAdYH | 0          | 0           |
| new_after_refill1   | mnYTcGF8vVuSAYeoxxNaFWX94QUFM92YZf | 0          | 0           |
| new_after_refill3   | n3qWzoBX24abCSq8hMMzmCb8dh5ePnhdmL | 0          | 0           |
| new_after_refill4   | n454R3cUohDkjPvfkpjMZysWjp6FkR8mfC | 0          | 0           |
| new_after_refill5   | n1zrFcgeDhcMP3PGyttiKviAW92TMrQf6r | 0          | 0           |
| new_after_refill6   | mffwdnxCEFgM2GoZD1Cd6yibPB43RewaVA | 0          | 0           |
| new_after_refill7   | n3eHVFDPbBHuq9CrNjirJ33GNTQuMnRZVy | 0          | 0           |
| new_after_refill8   | mkxLRzTvrEpPbFxkp4tnRGXQ6UKcKH25TA | 0          | 0           |
| new_after_refill9   | mtYvCVBtEayA5y6szGSrSgGwHQfe44Bgh2 | 0          | 0           |
| new_after_refill2   | mk9WFSTqgtTJbwb5g6pGpByTRbYKiSaX3w | 0          | 0           |
| fresh_after_refill  | n4UW9h3QvBRxdF9prmT8cP8p13Vm8VYJn9 | 0          | 0           |
| fresh_after_refill1 | mue63g5tHoM4qLQNBYzF7kSHbj8yvKfVYz | 0          | 0           |
| fresh_after_refill2 | mr99o4xUWMaVNmueFuqHMvZwRPxkLiQe8U | 0          | 0           |
| fresh_after_refill3 | mpo8FqEPH5iSakV9GZAhRwTrvHeJ98CWBv | 0          | 0           |
| fresh_after_refill4 | mwUJwJePvN3cNPaNB3LJkTLWenAXQV9ekw | 0          | 0           |
| fresh_after_refill5 | n4CLjrpKA5oLuEn63M3iWfGP7fyQvxkg6b | 0          | 0           |
| fresh_after_refill6 | myGqHjaG4BVnh51rh1g6w8mH2oJQZ3QCJP | 0          | 0           |
| fresh_after_refill7 | mwahMEM12h4m3nuq58MwetYDYPbYUr918i | 0          | 0           |
| fresh_after_refill8 | mtJoREAk9dRQQXR6KNAkXb1gYAd5LCkML9 | 0          | 0           |
| fresh_after_refill9 | mxTq1JuXLMZjaKA8LkdySk3sMwhRz4FEG2 | 0          | 0           |
| another_fresh       | mwLKewF56R8Rje2tAuQRJzDxS9XC1a1q23 | 0          | 0           |
| usage_checked       | myHPHhnncUaiFQGqN1FK2qFtwcfPE4Bkqc | 0          | 0           |
| nw_1                | mkbU3SGSMxt4tjVepPHGywWWJ7gYGrhoBG | 0          | 0           |
|                     | miHhMLaMWubq4Wx6SdTEqZcUHEGp8RKMZt | 0          | 0           |
|                     | n1WEs2vcEbXFQWe6GtsFgHfQmzavHVYMi4 | 0          | 0           |
|                     | mgLyx4nNSPjcimazmo5U2Z8khUiRZWq5hn | 0          | 1.0         |
|                     | mz6HENDiSFPYqsi7wDZ33p5jt7j9LFo6qr | 0          | 0           |
|                     | mvuMyLkLkYKtnq3DkPY6Xe91AYthEGcfym | 0          | 0           |
|                     | n4HXAq2XjJAt4cNRL9gpPjDS7r8hxmghVR | 0          | 0           |
|                     | mtMGkB2mVxzPNZCWHruRmG1SmsSJBtyCUD | 0          | 0           |
|                     | mixcqdAy4aFtuBV7uRebqBTjiFWCcS4UEp | 61600.0    | 0           |
|                     | mwutmXuJEzDvfs59nKY47Unm3M8cszxhf9 | 0          | 0           |
|                     | mxasNazVpdfTmjQcedfxtHQtb3NXKVGEzh | 0          | 0           |
|                     | mzP9BNtqjSNq2rJ58w8GpQNnez38NGagsU | 0          | 0           |
|                     | mguM6Xso6UEVT8uPYGnFMw9aPBZ2vtRKoH | 0          | 0           |
|                     | muhhYam1NXMSUqKx4uSrfjS7N7Fcz6j6Qx | 0          | 0           |
|                     | mmjaHuS1kTRMFJaTJqYMjdmF18n36eXHfD | 0          | 0           |
|                     | mkwJijAwqXNFcEBxuC93j3Kni43wzXVuik | 0          | 0           |
|                     | mgiYjQjREbtKcKKH1kFXo1ycWCfGB5aZd7 | 0          | 0           |
|                     | mpF9FcfSps5jPSG2vDdGVF883xNFez1UGT | 0          | 0           |
|                     | myKpK8viS9k8JSWEA9syitxLJUTqdSeSvr | 0          | 0           |
|                     | moDgh2HgPZKWpFrNesC4QVBzvBrV79p9pA | 0          | 0           |
|                     | mtvQZrb9BC8HFtXe87Ev5PEUeA1cSvP211 | 0          | 0           |
|                     | mqnBdCH2cuoSiQWjKoEZHfov77bzrpmtNX | 0          | 0           |
|                     | mxfu9BNbsaECVVBRiqzavcvp6xhB2cz9r8 | 0          | 0           |
+---------------------+------------------------------------+------------+-------------+
```
The `token list -s` output is:
```
+Standard PoB and dPoD decks (in this order):----------------------------------------------+-----------------+------------------------------------+---+--------+---+
| Local label           | ID                                                               | Global name     | Issuer                             | M | Conf.  | I |
+-----------------------+------------------------------------------------------------------+-----------------+------------------------------------+---+--------+---+
| ATTokenNewSpec2_local | fb93cce7aceb9f7fda228bc0c0c2eca8c56c09c1d846a04bd6a59cae2a895974 | ATTokenNewSpec2 | mvrm2HAoKqiCmeQiwKMuEFpeEn7rJEmpMz | 1 | 315007 | + |
| DTNewSpecv2_newl      | a2459e054ce0f600c90be458915af6bad36a6863a0ce0e33ab76086b514f765a | DTNewSpecv2     | mvrm2HAoKqiCmeQiwKMuEFpeEn7rJEmpMz | 1 | 314986 | + |
+-----------------------+------------------------------------------------------------------+-----------------+------------------------------------+---+--------+---+
```
I assume that the `a2459e054ce0f600c90be458915af6bad36a6863a0ce0e33ab76086b514f765a` deck is dPoD's.

However both my `token balances -w -t a2459e054ce0f600c90be458915af6bad36a6863a0ce0e33ab76086b514f765a` and `token balances -t a2459e054ce0f600c90be458915af6bad36a6863a0ce0e33ab76086b514f765a -w` outputs are:
```
{'mgLyx4nNSPjcimazmo5U2Z8khUiRZWq5hn': 1.0}
```
Thus the balance on the address `mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya` is not considered for some reason.

The address makes part of my wallet since I'm able to set it up as main with the `address set -a mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya` command.
_
As a side note I'd suggest us to consider specifying somehow in the `token list -s` output which deck is the dPoD's and which is the PoB's main, if possible.

UPDATE for this side note: maybe we can implement `token list -s -p` and `token list -s -b` for the user to be able to see the specific standard deck. For the moment the `-p` and `-b` flags are ignored when the `-s` flag is in place.
_
UPDATE
In the brand new Ubuntu18 docker container installation the address`mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya` is not even present in the `address list` output, but I'm able to set up it as main and am able to see both the coins and tokens on it ~and also to transfer the tokens~.
(I've copied the testnet folder from the host into container).
While I'm trying to transfer them I'm getting the following error:
```
  File "/usr/local/bin/pacli", line 11, in <module>
    load_entry_point('pacli==0.4.7.2+slm', 'console_scripts', 'pacli')()
  File "/usr/local/lib/python3.6/dist-packages/pacli/__main__.py", line 493, in main
    'checkpoint' : Checkpoint()
  File "/usr/local/lib/python3.6/dist-packages/fire/core.py", line 143, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/usr/local/lib/python3.6/dist-packages/fire/core.py", line 482, in _Fire
    target=component.__name__)
  File "/usr/local/lib/python3.6/dist-packages/fire/core.py", line 693, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/pacli/extended_classes.py", line 1307, in transfer
    return ei.run_command(self.__transfer, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/pacli/extended_interface.py", line 33, in run_command
    result = c(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/pacli/extended_classes.py", line 1339, in __transfer
    debug=debug
  File "/usr/local/lib/python3.6/dist-packages/pacli/extended_utils.py", line 388, in advanced_card_transfer
    return finalize_tx(issue_tx, verify=verify, sign=sign, send=send, quiet=quiet, confirm=confirm, debug=debug)
  File "/usr/local/lib/python3.6/dist-packages/pacli/extended_utils.py", line 228, in finalize_tx
    if reorg_check(quiet=quiet) and not ei.confirm_continuation():
  File "/usr/local/lib/python3.6/dist-packages/pacli/extended_checkpoints.py", line 221, in reorg_check
    stored_bhash = config["checkpoint"][str(last_height)]
UnboundLocalError: local variable 'last_height' referenced before assignment
```

**Comment by d5000** [comment]
> UPDATE for this side note: maybe we can implement token list -s -p and token list -s -b for the user to be able to see the specific standard deck. For the moment the -p and -b flags are ignored when the -s flag is in place.

Good idea, I implemented it this way and changed the table titles accordingly for these options in commit 10dc4a0.

Related to the side note:

My "workaround" was the "in this order" in the table title until now but I was already thinking about a better solution. The idea would be that the "standard" DPoD token would be stored, once the deck is initialized, with the local label "dPoD_standard" and the PoB as "PoB_standard" (or similar). Or even simply "dPoD" and "PoB", period. I also am considering to initialize these two decks automatically when the user first uses pacli, so the user isn't even tempted to give these names to other tokens. Have to think about possible implication though, but that's the idea to make that a bit more usable.

----

Checkpoint bug (last error message) has been fixed in commit 3aead6d.

----

Regarding your problem with the address: I'm seeing the correct balance with `token balances -o` so the token algorithm isn't the problem.

I can confirm it with your other wallet though, so I have to investigate more (`dumpprivkey` works with the address, this is the only method I recommand to really confirm that it's in the wallet).

**Comment by buhtignew** [comment]
Everything is solved except a new issue I've got on docker container that uses the original keyring of the Ubuntu18 version.
It's as if your commit https://github.com/slimcoin-project/pacli/commit/ee4ad6bffa235b574335403bdd2d0daaf35358b8 wasn't implemented in the docker's pacli version, but if I run ` git pull --ff-only` I'm getting `Already up to date.` and my `git checkout master` output is:
```
M       pacli/default_conf.py
Already on 'master'
Your branch is up to date with 'origin/master'.
```
I've been forced to install `cryptography==36.0.1` for this installation, because for the next version I'm receiving a warning:
```
/usr/local/lib/python3.6/dist-packages/secretstorage/util.py:23: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography. The next release of cryptography will remove support for Python 3.6.
  from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
```
My `token balances a2459e054ce0f600c90be458915af6bad36a6863a0ce0e33ab76086b514f765a -o` output is:
```
    'mmBwLVyqZWuUDuTpmmpzBvT41azKBevZhh': 41.37,
    'mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya': 7.0,
    'mgLyx4nNSPjcimazmo5U2Z8khUiRZWq5hn': 1.0,
    'mtMGkB2mVxzPNZCWHruRmG1SmsSJBtyCUD': 1.0,
    'mzP9BNtqjSNq2rJ58w8GpQNnez38NGagsU': 1.0
```
However my `token balances -w -t a2459e054ce0f600c90be458915af6bad36a6863a0ce0e33ab76086b514f765a` output is:
```
    'mgLyx4nNSPjcimazmo5U2Z8khUiRZWq5hn': 1.0,
    'mtMGkB2mVxzPNZCWHruRmG1SmsSJBtyCUD': 1.0,
    'mzP9BNtqjSNq2rJ58w8GpQNnez38NGagsU': 1.0
```
My `pacli address list | grep mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya` is empty.
I'm able to set the address up as main with `address set -a mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya` command and then to see it's balances with `token balances -t a2459e054ce0f600c90be458915af6bad36a6863a0ce0e33ab76086b514f765a`, `token balances -t fb93cce7aceb9f7fda228bc0c0c2eca8c56c09c1d846a04bd6a59cae2a895974` and to move tokens with `token transfer fb93cce7aceb9f7fda228bc0c0c2eca8c56c09c1d846a04bd6a59cae2a895974 mijeUJNpC1Kqidne8aq7KbXDb5fApSWd39 1` command for instance.

However I'm not sure this info is appropriate for this issue or it should be discussed in #95 and also I'm not sure whether we should work at this at all since I'm feeling there are compatibilities issues in Ubuntu18.
By other hand without Ubuntu18 it would be difficult to have slimcoind compiled.

**Comment by d5000** [comment]
The token balances output looks like the output I got before the bug from yesterday was fixed (basically, both named addresses, mmBwLVyqZWuUDuTpmmpzBvT41azKBevZhh and mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya were excluded).

Are you tied to Ubuntu or can you also use Debian? As I wrote before I was able to compile slimcoind in Debian 11, and I can try to do it again and then reproduce the steps needed to you.

**Comment by buhtignew** [comment]
> The token balances output looks like the output I got before the bug from yesterday was fixed (basically, both named addresses, mmBwLVyqZWuUDuTpmmpzBvT41azKBevZhh and mtmKbmMHdcXXepscuNrwAfjsKLeZzdZdya were excluded).

Maybe my Ubuntu installation for some reason considers its `pacli` up to date while it's not? What test can I do locally to see whether you recent patch is in place?

> Are you tied to Ubuntu or can you also use Debian? 

I'm not particularly fond of Ubuntu18. The reason I'm using that particular distribution and that version is that it's the most recent one that let compile slimcoind smoothly, even without warnings.
And it's still available for docker and by many VPS providers.
So for example I was able compile slimcoind statically inside of Ubuntu18 docker container to use it later on the more recent host OS just by copying the appropriate boost library files from the Ubuntu18 container together with slimcoind.

I've tried to compile on Debian11 but it doesn't compile without editing the code probably because Debian11 uses a more recent version of boost (1.74), while Ubuntu18 has still 1.65. And even by replacing boost 1.74 with 1.65 I was not able to compile slimcoind there. 
I'm confident it would be possible to solve the issue, but having found a trick to use slimcoind on the more recent versions as described above I'm not sure it worth dedicating time to find the way to compile it on Debian11 as well.

The most recent version of Debian that let compile slimcoind smoothly, but still with some warning is 10. It's considered oldoldstable and is more recent than Ubuntu18 and uses more recent version of boost (1.67). So it can be an option in the future when we won't be able to find the way to use Ubuntu18 anymore, but I have an impression that it would be more difficult to find Debian10 by VPS providers.

However since I had Ubuntu18 container already available locally and on the server I thought it would make sense trying it with our pacli code. Which, seeing the number of issues we are witnessing, wasn't probably a good idea, unless those issues are helping our pacli code to improve.

In any case my main concern is to find a solution to provide to our users that would let them smoothly install and use slimcoind+pacli.

But probably we should be testing them both on more recent OSs.

Another concern I have is that since our core code is becoming more and more old we are risking to reach the point when we won't be able to find an OS anymore to implement and to test PPC commits in the order they were pushed, if we opt to follow that way to move our core development.

What do you think about what I've expressed in the last three sections?

**Comment by d5000** [comment]
> But probably we should be looking for a solution for the more recent OSs.
Agree with that line.

> Another concern I have is that since our core code is becoming more and more old we are risking to reach the point when we won't be able to find an OS anymore to implement and to test PPC commits in the order they were pushed

We already talked about that and I think this is the wrong strategy. It's better to instead take a new PPC version and apply the slimcoind changes to it. That's what I was doing for a time last year but then I ran into a problem and wanted to first focus on the dPoD/PoB tokens solution first. My last version of that edited PPC version was able to connect to SLM testnet but did not download blocks correctly. It's not unsolvable for me, only that the pacli/peerassets testing takes away all the time I could otherwise dedicate to solving that issue.
